lib/Dialect/Moore/Transforms/LowerConcatRef.cpp

```diff
@@ -0,0 +1,128 @@
+//===- LowerConcatRef.cpp - moore.concat_ref lowering ---------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines the LowerConcatRef pass.
+// It's used to disassemble the moore.concat_ref. Which is tricky to lower
+// directly. For example, disassemble "{a, b} = c" onto "a = c[7:3]"
+// and "b = c[2:0]".
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Dialect/Moore/MooreOps.h"
+#include "circt/Dialect/Moore/MoorePasses.h"
+#include "mlir/IR/ValueRange.h"
+#include "mlir/Support/LogicalResult.h"
+#include "mlir/Transforms/DialectConversion.h"
+
+namespace circt {
+namespace moore {
+#define GEN_PASS_DEF_LOWERCONCATREF
+#include "circt/Dialect/Moore/MoorePasses.h.inc"
+} // namespace moore
+} // namespace circt
+
+using namespace circt;
+using namespace moore;
+using namespace mlir;
+
+namespace {
+
+// struct ConcatRefLowering : public OpConversionPattern<ConcatRefOp> {
+//   using OpConversionPattern::OpConversionPattern;
+
+//   LogicalResult
+//   matchAndRewrite(ConcatRefOp op, OpAdaptor adaptor,
+//                   ConversionPatternRewriter &rewriter) const final {
+//     auto *user = *op->getUsers().begin();
+
+//     return success();
+//   }
+// };
+
+// template<typename OpTy>
+struct ConcatRefLowering : public OpConversionPattern<ConcatRefOp> {
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(ConcatRefOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const final {
+    if (auto *user = *op->getUsers().begin()) {
```

Comment:
I think you have to all users which include non-assignments (e.g. concat ref).

You can flatten concat_ref by flattening. 

`concat_ref(concat_ref(a, b), c) -> concat_ref(a, b, c)`.

Reply:
Exactly! I have thought about this you mentioned. But for the IRs, merge them
```
%0 = moore.concat_ref %a, %a : (!moore.ref<i32>, !moore.ref<i32>) -> <i64>
%1 = moore.concat_ref %a, %b : (!moore.ref<i32>, !moore.ref<i32>) -> <i64>
%2 = moore.concat_ref %0, %1 : (!moore.ref<i64>, !moore.ref<i64>) -> <i128>
```
into one op like `%0 = moore.concat_ref %a, %a, %a, %b : ...` looks like difficult.
When I visit `%0`, I must handle it immediately, right? If I don't handle it immediately. I must wait until visit `%2`, then recursively traverse its operands(`%0` and `%1`), and create a new `concat_ref`.

Reply:
> into one op like %0 = moore.concat_ref %a, %a, %a, %b : ... looks like difficult.

You already handled variadic operands so I think it works without any problem?

Reply:
Maybe I can create a container like SmallVector, and then traverse the operands of `concat_ref`, and then collect them.